domains/client/domain-executor/src/domain_bundle_proposer.rs

```diff
@@ -161,29 +162,36 @@ where
 
         let header_number = to_number_primitive(header_number);
 
-        // Ideally, the receipt of current block will be included in the next block, i.e., no
-        // missing receipts.
-        let receipts = if header_number == head_receipt_number + 1 {
-            let block_hash = self.client.hash(header_number.into())?.ok_or_else(|| {
-                sp_blockchain::Error::Backend(
-                    format!("Hash for Block {header_number:?} not found",),
-                )
-            })?;
-            vec![load_receipt(block_hash)?]
-        } else {
-            // Receipts for some previous blocks are missing.
-            let max_allowed = (head_receipt_number + max_drift).min(header_number);
-
-            let mut to_send = head_receipt_number + 1;
-            let mut receipts = Vec::with_capacity((max_allowed - to_send + 1) as usize);
-            while to_send <= max_allowed {
-                let block_hash = self.client.hash(to_send.into())?.ok_or_else(|| {
-                    sp_blockchain::Error::Backend(format!("Hash for Block {to_send:?} not found"))
+        let receipts = match header_number.cmp(&(head_receipt_number + 1)) {
+            // Ideally, the receipt of current block will be included in the next block, i.e., no
+            // missing receipts.
+            Ordering::Equal => {
+                let block_hash = self.client.hash(header_number.into())?.ok_or_else(|| {
+                    sp_blockchain::Error::Backend(format!(
+                        "Hash for Block {header_number:?} not found",
+                    ))
                 })?;
-                receipts.push(load_receipt(block_hash)?);
-                to_send += 1;
+                vec![load_receipt(block_hash)?]
+            }
+            // Receipts for some previous blocks are missing.
+            Ordering::Greater => {
+                let max_allowed = (head_receipt_number + max_drift).min(header_number);
+                let mut to_send = head_receipt_number + 1;
+                let mut receipts = Vec::with_capacity((max_allowed + 1 - to_send) as usize);
```

Comment:
Although we only observed the case `max_allowed == to_send`, with the abovementioned scenario (i.e the executor is even lagging behind the receipt chain on its parent chain) it is possible that `max_allowed < to_send`, and IMHO having a dedicated branch for each possible case can make the logic more clear and readable.

Reply:
Well, with the check beforehand, it's theoretically not possible to reach this branch, that's why `unreabable!` statement could stand. If you insist keeping it, please return an error instead of killing the program, it doesn't seem to have to be his critical.

Reply:
More thoughts on this: in the case of `head_receipt_number == header_number`, changing this line to `max_allowed + 1 - to_send` also doesn't help because this branch will return empty receipts.

In the first commit, I introduce this line of change out of the concerns about `max_drift` (I have a quick look and didn't find any constraint about `max_drift != 0`), if `max_drift == 0` then it will be possible that:
- max_allowed = head_receipt_number
- to_send = head_receipt_number + 1

Thus `max_allowed - to_send + 1` will underflow.

Reply:
As a matter of fact, we can rewrite it completely for better readability, which is much simpler and avoids all these hassles. It's clear that the first receipt is `head_receipt_number + 1`, new receipts after that can be added continuously until we can't.

```Rust
        let mut receipts = Vec::new();
        let mut to_send = head_receipt_number + 1;
        loop {
            let block_hash = self.client.hash(to_send.into())?.ok_or_else(|| {
                sp_blockchain::Error::Backend(format!("Hash for Block {to_send:?} not found"))
            })?;
            receipts.push(load_receipt(block_hash)?);
            to_send += 1;

            if to_send > header_number || to_send > head_receipt_number + max_drift {
                break;
            }
        }
```

Reply:
The above proposal will at least include one receipt, but on genesis, it will return an error due to:
```rust
let block_hash = self.client.hash(to_send.into())?.ok_or_else(|| {
    sp_blockchain::Error::Backend(format!("Hash for Block {to_send:?} not found"))
})?;
```
perhaps we should just use this branch with `Vec:: with_capacity` replace to `Vec::new`?
https://github.com/subspace/subspace/blob/aa2133611127f6fc84c281d26491cbb60a4a7d41/domains/client/domain-executor/src/domain_bundle_proposer.rs#L174-L186

Reply:
>on genesis, it will return an error due to:

This is wrong because this function is invoked on non-genesis.

https://github.com/subspace/subspace/blob/d5cf9e6295f4535085698359337c83440a1d76f7/domains/client/domain-executor/src/domain_bundle_proposer.rs#L106-L110

Reply:
>It may invoke on the genesis of the domain chain.

Please elaborate, I don't get it.

Reply:
I mean when the primary chain's best number is 1, the domain's best number could be 0, i.e the primary chain is ahead of the domain by one block.

Reply:
I think this should handled before attempting to produce a bundle. In order to produce a bundle, the executor should be fully synced:

- The corresponding primary node is fully synced and the executor should have processed the latest primary block.
- The executor has already processed the block the best receipt on chain points to.

So the executor should skip this slot in this case and try to process the primary block 1 in highest priority.

Regarding this new definition of fully synced executor, Ved had a discussion with me yesterday that we agree there should be something like `DomainSyncOracle` indicating whether it's domain-wise fully synced, which is a reusable component used both on the bundle production and relayer side. But it doesn't have to be completely done in this PR, we can file a new issue for it. cc @vedhavyas